isomerpages/isomercms-backend
# Feat/site launch

## Problem

Automate site launch process

Partially Addresses #516 

Steps to test locally: 

1. Run database migrations by running `npx sequelize-cli db:migrate`
2. Populate the `sites` table database with at least one site
3. Run through the steps [here](https://github.com/isomerpages/isomercms-backend/tree/feat/add-serverless#running-locally) to start local stack and step functions 
4. Create incoming and outgoing queues by running: `aws --endpoint-url=<URL_OF_LOCALSTACK> sqs create-queue --queue-name incomingQueue` and `aws --endpoint-url=<URL_OF_LOCALSTACK> sqs create-queue --queue-name outgoingQueue`
5. Update `INCOMING_QUEUE_URL` and `OUTGOING_QUEUE_URL` in .env file
6. Start backend server 
7. Submit request to launch site via FormSG
8. Check to see if the database tables have been updated
9. Check Amplify console to see domain being added 

## Deploy Notes

This PR involves addition of two tables in database. This would require DB migrations when merging to prod/staging. As discussed in the eng sync, there was a presence of a indeterministic bug (similar issue documented [here](https://github.com/aws/aws-sdk-js/issues/1453) and [here](https://github.com/aws/aws-sdk-js/issues/2793)) which I have tried to fix this in the commit [here](https://github.com/isomerpages/isomercms-backend/pull/517/commits/e9f3e06471119c35c7e2ddf18a24cd2ea86b2e17), but the issue is still not fixed :( . This is slightly scary as wrapping the error exists outside of try/catch block, so in the worst case, this will crash the server. Another dirty solution to be to modify the aws sdk as described [here](https://github.com/aws/aws-sdk-js/issues/2700). Running this on staging for a prolonged period of time (maybe a day/over the weekend) with some manual integration tests might be important before merging to prod. ._. 

**New environment variables**:

- `INCOMING_QUEUE_URL` : incoming queue's URL 
- `OUTGOING_QUEUE_URL` : outgoing queue's URL 

TODO before merging to prod: 
[ ] Remove `isDev` flag [here](https://github.com/isomerpages/isomercms-backend/blob/feat/site-launch/src/routes/formsgSiteLaunch.ts#L67) 
[ ] Fix network issue in `Request.HTTP_DATA` in aws SDK
